File: table-order/backend/app/controllers/order.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies import get_current_admin, get_current_user, require_role
from app.exceptions import NotFoundError
from app.models.order import Order
from app.schemas.order import OrderCreate, OrderResponse, OrderStatusUpdate
from app.services import order_service

router = APIRouter()


@router.post("", response_model=OrderResponse)
def create_order(req: OrderCreate, db: Session = Depends(get_db), _user: dict = Depends(get_current_user)):
    items = [{"menu_id": i.menu_id, "quantity": i.quantity} for i in req.items]
    return order_service.create_order(db, req.store_id, req.table_id, req.session_id, items)


@router.get("", response_model=list[OrderResponse])
def list_orders(
    store_id: int = Query(...),
    table_id: int | None = Query(None),
    session_id: int | None = Query(None),
    db: Session = Depends(get_db),
    _user: dict = Depends(get_current_user),
):
    q = db.query(Order).filter(Order.store_id == store_id, Order.is_deleted.is_(False))
    if table_id:
        q = q.filter(Order.table_id == table_id)
    if session_id:
        q = q.filter(Order.session_id == session_id)
    return q.order_by(Order.created_at.desc()).all()


@router.get("/{order_id}", response_model=OrderResponse)
def get_order(order_id: int, db: Session = Depends(get_db), _user: dict = Depends(get_current_user)):
    order = db.query(Order).filter(Order.id == order_id, Order.is_deleted.is_(False)).first()
    if not order:
        raise NotFoundError("Order not found")
    return order


@router.patch("/{order_id}/status", response_model=OrderResponse)
def update_status(order_id: int, req: OrderStatusUpdate, db: Session = Depends(get_db), _user: dict = Depends(get_current_admin)):
    return order_service.update_order_status(db, order_id, req.status, _user["store_id"])


@router.delete("/{order_id}")
def delete_order(order_id: int, db: Session = Depends(get_db), _user: dict = Depends(require_role("manager"))):
    order_service.delete_order(db, order_id, _user["store_id"])
    return {"success": True}

File: table-order/frontend/src/actions/order.ts
"use server";
import { apiFetch } from "@/lib/api";
import type { Order } from "@/lib/types";

export async function createOrder(token: string, data: { store_id: number; table_id: number; session_id?: number; items: { menu_id: number; quantity: number }[] }): Promise<Order> {
  return apiFetch<Order>("/api/orders", { method: "POST", body: JSON.stringify(data), token });
}

export async function fetchOrders(token: string, storeId: number, tableId?: number, sessionId?: number): Promise<Order[]> {
  let url = `/api/orders?store_id=${storeId}`;
  if (tableId) url += `&table_id=${tableId}`;
  if (sessionId) url += `&session_id=${sessionId}`;
  return apiFetch<Order[]>(url, { token });
}

export async function updateOrderStatus(token: string, orderId: number, status: string): Promise<Order> {
  return apiFetch<Order>(`/api/orders/${orderId}/status`, { method: "PATCH", body: JSON.stringify({ status }), token });
}

export async function deleteOrder(token: string, orderId: number): Promise<void> {
  await apiFetch(`/api/orders/${orderId}`, { method: "DELETE", token });
}
